### ISSUE TITLE
pos_weight for dealing with imbalanced dataset

### ISSUE THREAD

**Issue by hulkds** [ultralytics/ultralytics#8578]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Add a weight of positive examples to be broadcasted with target (for Detection).

### Use case

Based on [torch.nn.BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html), add pos_weight parameter can help when working with imbalanced dataset.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by hulkds** [comment]
PR: https://github.com/ultralytics/ultralytics/pull/8557

**Comment by glenn-jocher** [comment]
@hulkds thank you for bringing up the idea of incorporating `pos_weight` for handling imbalanced datasets in YOLOv8 and for the PR! 🚀 This is indeed a valuable addition, especially for those working with datasets where the class distribution is skewed. 

Your suggestion to leverage `torch.nn.BCEWithLogitsLoss`'s `pos_weight` parameter aligns well with our goal to enhance model performance across a variety of scenarios. We'll review the PR at https://github.com/ultralytics/ultralytics/pull/8557 closely and provide feedback or merge it after thorough testing.

For anyone interested, this feature allows you to adjust the weight of positive examples dynamically, which can be particularly useful when dealing with classes that are underrepresented in your dataset. Here's a quick snippet on how you might adjust the loss function:

```python
import torch.nn as nn

# Assuming 'pos_weight' is calculated based on your dataset's class distribution
loss_fn = nn.BCEWithLogitsLoss(pos_weight=pos_weight)
```

We appreciate your contribution and look forward to more collaborations! 🙌

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082

**Comment by Lalita793** [comment]
How to apply pose_weight in classification loss for unbalance datasets . How to customize it in yolov12

**Comment by Lalita793** [comment]
After applying pose_weight how we can verify it the customize loss function is apply or not

**Comment by glenn-jocher** [comment]
@Lalita793 It sounds like you meant pos_weight, but for classification Ultralytics uses CrossEntropy (not BCE), so you’d apply per-class weights via the `weight` tensor rather than `pos_weight`; you can plug in a custom criterion like this and assign it before training (recommend YOLO11 over YOLO12 for stability):

```python
from ultralytics import YOLO
import torch
import torch.nn.functional as F

class WeightedCE:
    def __init__(self, class_weights):
        self.class_weights = torch.as_tensor(class_weights, dtype=torch.float32)
        self._logged = False

    def __call__(self, preds, batch):
        preds = preds[1] if isinstance(preds, (list, tuple)) else preds
        w = self.class_weights.to(preds.device)
        if not self._logged:  # quick one-time verification
            print(f"Using class weights: {w.tolist()}")
            self._logged = True
        loss = F.cross_entropy(preds, batch["cls"], weight=w, reduction="mean")
        return loss, loss.detach()

# usage
model = YOLO("yolo11n-cls.pt")
model.model.criterion = WeightedCE([w0, w1, ...])  # one weight per class
model.train(data="path/to/data.yaml", epochs=..., imgsz=...)
```

You can confirm it’s active by checking `type(model.model.criterion)` and the one-time print; for more on the built-in classification loss see the [v8ClassificationLoss reference](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.v8ClassificationLoss). This thread is closed; for follow-up please open a new issue with `yolo checks` output and a minimal reproducible example for visibility.

**Comment by Lalita793** [comment]
How to apply pose weight for class weight tuning in classification loss but bbox loss and dfl loss will unchanged after that we calculate total loss

**Comment by glenn-jocher** [comment]
Assuming you mean pos_weight: for detection you can weight only the classification term by setting `BCEWithLogitsLoss.pos_weight` on the criterion; the box and DFL terms remain unchanged since they’re computed separately in `v8DetectionLoss` (see the BCE use in the [loss reference](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.v8OBBLoss) and how the criterion is created in [DetectionModel.init_criterion](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.DetectionModel.init_criterion)); example:

```python
from ultralytics import YOLO
import torch

pos_weight = torch.tensor([w0, w1, ...], dtype=torch.float32)  # one value per class

def set_cls_pos_weight(trainer):
    # ensure criterion exists, then set pos_weight on its BCE
    if getattr(trainer.model, "criterion", None) is None:
        trainer.model.criterion = trainer.model.init_criterion()
    bce = getattr(trainer.model.criterion, "bce", None)
    if bce:
        bce.pos_weight = pos_weight.to(trainer.device)
        print("pos_weight set:", bce.pos_weight.tolist())

model = YOLO("yolo11n.pt")  # recommend YOLO11 for stability
model.add_callback("on_train_start", set_cls_pos_weight)
model.train(data="data.yaml", epochs=..., imgsz=...)
```

This issue is closed; if you need further help please open a new issue with your `yolo checks` output and a minimal reproducible example.
